software/controller/lib/hal/spi_chain.tpp

```diff
@@ -0,0 +1,235 @@
+/* Copyright 2022, RespiraWorks
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+////////////////////////////////////////////////////////////////////////////////
+/// WARNING: Do not include this file as header, nor compile it as a source! ///
+///                                                                          ///
+/// This is a template implementation file.                                  ///
+/// It is included by the corresponding header.                              ///
+////////////////////////////////////////////////////////////////////////////////
+
+template<size_t MaxSlaves>
+DaisyChain<MaxSlaves>::DaisyChain(Base spi, DMA::Base dma, const char* name, const char* help_supplement)
+   : Channel(spi, dma)
+  {
+    queueing_errors_.prepend_name(name);
+    queueing_errors_.append_help(help_supplement);
+   }
+
+template<size_t MaxSlaves>
+void DaisyChain<MaxSlaves>::Initialize(uint8_t null_command, uint8_t reset_command,
+                                       GPIO::Port clock_port, uint8_t clock_pin,
+                                       GPIO::Port miso_port, uint8_t miso_pin,
+                                       GPIO::Port mosi_port, uint8_t mosi_pin,
+                                       GPIO::Port chip_select_port, uint8_t chip_select_pin,
+                                       GPIO::Port reset_port, uint8_t reset_pin,
+                                       uint8_t word_size, Bitrate bitrate) {
+  Channel::Initialize(clock_port, clock_pin, miso_port, miso_pin, mosi_port, mosi_pin,
+                      chip_select_port, chip_select_pin, reset_port, reset_pin, word_size,
+                      bitrate, /*rx_interrupts_enabled=*/true,
+                      /*tx_interrupts_enabled=*/false, /*rx_listener=*/this);
+  null_command_=null_command;
+  ProbeSlaves(null_command, reset_command);
+}
+
+template<size_t MaxSlaves>
```

Comment:
I wonder if some of these functions could not be broken down into smaller ones? This might help with somehow isolating the general functions from the "mock-worthy" ones?

Reply:
done (to some extent)